lib/src/stock_response_extensions.dart

```diff
@@ -3,6 +3,67 @@ import 'package:stock/src/stock_response.dart';
 
 /// Useful [StockResponse] extensions.
 extension StockResponseExtensions<T> on StockResponse<T> {
+  /// Invoke [onData] if the response is successful, [onLoading] if the response
+  /// is loading, and [onError] if the response is an error.
+  E map<E>({
+    required E Function(StockResponseLoading<T> value) onLoading,
+    required E Function(StockResponseData<T> value) onData,
+    required E Function(StockResponseError<T> value) onError,
+  }) {
+    if (this is StockResponseLoading<T>) {
+      return onLoading(this as StockResponseLoading<T>);
+    } else if (this is StockResponseData<T>) {
+      return onData(this as StockResponseData<T>);
+    } else if (this is StockResponseError<T>) {
+      return onError(this as StockResponseError<T>);
+    } else {
+      throw StockError('Unknown StockResponse type: $this');
+    }
+  }
+
+  /// Invoke [onData] or [orElse] as fallback if the response is successful,
+  /// [onLoading] or [orElse] as fallback if the response is loading, and
+  /// [onError] or [orElse] as fallback if the response is an error.
+  E maybeMap<E>({
+    E Function(StockResponseLoading<T> value)? onLoading,
+    E Function(StockResponseData<T> value)? onData,
+    E Function(StockResponseError<T> value)? onError,
+    required E Function() orElse,
+  }) =>
+      map(
+        onLoading: onLoading ?? (_) => orElse(),
+        onData: onData ?? (_) => orElse(),
+        onError: onError ?? (_) => orElse(),
+      );
+
+  /// Invoke [onData] if the response is successful, [onLoading] if the response
+  /// is loading, and [onError] if the response is an error.
+  E when<E>({
```

Comment:
What do you think of adding the origin? For example it could be useful to "stop the loading indicator"

Reply:
Yeah, that would be useful.